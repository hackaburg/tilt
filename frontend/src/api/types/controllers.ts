export * from "../../../../backend/src/controllers/api";
export * from "../../../../backend/src/controllers/settings-controller";
export * from "../../../../backend/src/controllers/users-controller";
export * from "../../../../backend/src/controllers/application-controller";
