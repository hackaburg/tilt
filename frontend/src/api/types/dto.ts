export * from "../../../../backend/src/controllers/dto";
