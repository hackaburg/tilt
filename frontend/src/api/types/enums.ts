export { UserRole } from "../../../../backend/src/entities/user-role";
export { QuestionType } from "../../../../backend/src/entities/question-type";
