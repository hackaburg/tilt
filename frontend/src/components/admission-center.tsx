import * as React from "react";
import { Heading } from "./headings";

/**
 * One table to admit them all.
 */
export const AdmissionCenter = () => {
  return (
    <>
      <Heading>Admission Center</Heading>
    </>
  );
};
