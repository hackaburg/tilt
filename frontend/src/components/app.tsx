import * as React from "react";
import { useEffect } from "react";
import { RouteComponentProps, withRouter } from "react-router";
import { defaultThemeColor } from "../config";
import { useLoginContext } from "../contexts/login-context";
import { useSettingsContext } from "../contexts/settings-context";
import { Routes } from "../routes";
import { ThemeProvider } from "../theme";
import { ErrorBoundary } from "./base/error-boundary";
import { LazyAuthenticatedRouter } from "./routers/lazy-authenticated-router";
import { UnauthenticatedRouter } from "./routers/unauthenticated-router";

interface IAppProps extends RouteComponentProps<any> {}

/**
 * The main app component.
 */
export const App = ({ history, location }: IAppProps) => {
  const { isLoggedIn } = useLoginContext();
  const { pathname } = location;

  useEffect(() => {
    if (isLoggedIn) {
      if (pathname === Routes.Login) {
        history.push(Routes.Status);
      }
    } else {
      if (pathname !== Routes.VerifyEmail) {
        history.push(Routes.Login);
      }
    }
  }, [isLoggedIn, pathname]);

  let theme = {
    colorGradientEnd: defaultThemeColor,
    colorGradientStart: defaultThemeColor,
    colorLink: defaultThemeColor,
    colorLinkHover: defaultThemeColor,
  };

  const { settings } = useSettingsContext();

  if (settings) {
    theme = {
      colorGradientEnd: settings.frontend.colorGradientEnd,
      colorGradientStart: settings.frontend.colorGradientStart,
      colorLink: settings.frontend.colorLink,
      colorLinkHover: settings.frontend.colorLinkHover,
    };
  }

  const router = isLoggedIn ? (
    <LazyAuthenticatedRouter />
  ) : (
    <UnauthenticatedRouter />
  );

  return (
    <ThemeProvider values={theme}>
      <ErrorBoundary>{router}</ErrorBoundary>
    </ThemeProvider>
  );
};

/**
 * The main app component, connected to react-router.
 */
export const RoutedApp = withRouter(App);
