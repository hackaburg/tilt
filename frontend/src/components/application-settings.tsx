import * as React from "react";
import { useDebouncedCallback } from "use-debounce";
import { IApplicationSettings } from "../../../types/settings";
import { debounceDuration } from "../config";
import { useSettingsContext } from "../contexts/settings-context";
import { FormEditor } from "./form-editor";
import { Col, Row } from "./grid";
import { Subheading } from "./headings";
import { Message } from "./message";
import { StatefulTextInput, TextInputType } from "./text-input";

/**
 * Settings to configure the application users have to fill out.
 */
export const ApplicationSettings = () => {
  const { settings, updateSettings, updateError } = useSettingsContext();

  const [debouncedHandleUpdateApplicationSettings] = useDebouncedCallback(
    (field: keyof IApplicationSettings, value: any) => {
      updateSettings({
        ...settings,
        application: {
          ...settings.application,
          [field]: value,
        },
      });
    },
    debounceDuration,
    [settings, updateSettings],
  );

  return (
    <>
      <Subheading>Application</Subheading>
      {updateError && (
        <Message error>
          <b>Error:</b> {updateError.message}
        </Message>
      )}

      <p>
        An application is divided into two parts: the application and the
        confirmation phase. Once you accept applications, the users will be
        moved to the confirmation queue, where they'll need to fill out the
        remaining questions. If you add questions to the first phase after users
        submitted the first answers, tilt will ask these new questions in the
        confirmation phase.
      </p>
      <Row>
        <Col percent={33}>
          <StatefulTextInput
            initialValue={settings.application.hoursToConfirm}
            onChange={(time) =>
              debouncedHandleUpdateApplicationSettings("hoursToConfirm", time)
            }
            type={TextInputType.Number}
            min={1}
            title="Hours to confirm"
            placeholder="keep it fair, e.g. 240 for 10 days"
          />
        </Col>

        <Col percent={33}>
          <StatefulTextInput
            initialValue={settings.application.allowProfileFormFrom}
            onChange={(timestring) =>
              debouncedHandleUpdateApplicationSettings(
                "allowProfileFormFrom",
                timestring,
              )
            }
            title="Open registration on"
            placeholder="1970-01-01 00:00:00"
          />
        </Col>

        <Col percent={33}>
          <StatefulTextInput
            initialValue={settings.application.allowProfileFormUntil}
            onChange={(timestring) =>
              debouncedHandleUpdateApplicationSettings(
                "allowProfileFormUntil",
                timestring,
              )
            }
            title="Close registration on"
            placeholder="1970-01-01 00:00:00"
          />
        </Col>
      </Row>

      <p>
        Use the add button to add new questions and the edit button in the top
        right of each question to modify them. You may use Markdown syntax in
        the description, but please keep it short.
      </p>
      <p>
        Questions can have reference names, which you can use to conditionally
        show other questions. For instance, if you have a question, whether
        someone is a student, you could give that question the reference name
        "student" and modify subsequent questions by referencing the user's
        answer to that question.
      </p>

      <FormEditor
        initialForm={settings.application.profileForm}
        onFormChange={(form) =>
          debouncedHandleUpdateApplicationSettings("profileForm", form)
        }
      />

      <FormEditor
        initialForm={settings.application.confirmationForm}
        onFormChange={(form) =>
          debouncedHandleUpdateApplicationSettings("confirmationForm", form)
        }
      />
    </>
  );
};
