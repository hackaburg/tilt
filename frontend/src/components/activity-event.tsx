import * as React from "react";
import { useState } from "react";
import styled from "styled-components";
import { ActivityType, IActivity } from "../../../types/activity";
import { borderRadius, transitionDuration } from "../config";
import { Nullable } from "../state";
import { IThemeProps } from "../theme";
import { dateToString } from "../util";
import { DiffEditor } from "./diff-editor";

const Container = styled.div`
  margin-bottom: 1rem;
`;

interface IBodyProps {
  shown: boolean;
}

const Body = styled.div<IBodyProps>`
  opacity: 0;
  padding: 0rem 1rem;
  height: 0rem;
  overflow: hidden;

  ${({ shown }) =>
    shown &&
    `
    opacity: 1;
    height: calc(40vh + 2rem);
    padding: 1rem;
  `}

  border-radius: ${borderRadius};
  box-shadow: 0px 5px 10px rgba(0, 0, 0, 0.05);
  transition-property: height, opacity, padding;
  transition-duration: ${transitionDuration};
`;

interface ITitleProps {
  clickable: boolean;
}

const Title = styled.h3<ITitleProps>`
  margin: 0rem;
  font-weight: normal;
  font-size: 0.9rem;

  ${({ clickable }) =>
    clickable &&
    `
    cursor: pointer;
  `}
`;

const Accent = styled.span`
  color: ${({ theme }: IThemeProps) => theme.colorGradientEnd};
`;

const Time = styled.div`
  margin-bottom: 1rem;
  font-size: 0.75;
  opacity: 0.5;
`;

interface IClickIndicatorProps {
  clicked: boolean;
}

const ClickIndicator = styled.button<IClickIndicatorProps>`
  display: inline-block;
  margin: 0rem 0.5rem;
  border: none;

  color: currentColor;
  background-color: transparent;
  cursor: pointer;

  transition-property: transform;
  transition-duration: ${transitionDuration};

  ${({ clicked }) =>
    clicked &&
    `
    transform: rotateZ(90deg);
  `}
`;

const getActivityText = (type: ActivityType) => {
  switch (type) {
    case ActivityType.Signup:
      return "Signup";

    case ActivityType.EmailVerified:
      return "E-mail verified";

    case ActivityType.SettingsUpdate:
      return "Settings updated";
  }
};

interface IActivityEventProps {
  event: IActivity;
}

/**
 * An activity event.
 */
export const ActivityEvent = ({ event }: IActivityEventProps) => {
  const [showBody, setShowBody] = useState(false);
  let body: Nullable<JSX.Element> = null;

  switch (event.data.type) {
    case ActivityType.SettingsUpdate:
      body = (
        <DiffEditor
          language="json"
          left={event.data.previous}
          right={event.data.next}
        />
      );
      break;
  }

  return (
    <Container>
      <Title onClick={() => setShowBody((value) => !value)} clickable={!!body}>
        {getActivityText(event.data.type)} by{" "}
        <Accent>{event.user.email}</Accent>
        {body && <ClickIndicator clicked={showBody}>&#9658;</ClickIndicator>}
      </Title>

      <Time>{dateToString(new Date(event.timestamp))}</Time>

      {body && <Body shown={showBody}>{showBody && body}</Body>}
    </Container>
  );
};
