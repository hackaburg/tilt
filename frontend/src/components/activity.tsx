import * as React from "react";
import { useEffect } from "react";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import { IActivity } from "../../../types/activity";
import { fetchActivities } from "../actions/activity";
import { IState } from "../state";
import { groupByMonth } from "../util";
import { ActivityEvent } from "./activity-event";
import { Heading, Subheading } from "./headings";
import { Placeholder } from "./placeholder";

interface IActivityProps {
  activity: IActivity[] | null;
  dispatchFetchActivities: typeof fetchActivities;
}

/**
 * An overview over activities in tilt.
 */
export const Activity = ({
  activity,
  dispatchFetchActivities,
}: IActivityProps) => {
  useEffect(() => {
    if (!activity) {
      dispatchFetchActivities();
    }
  }, []);

  const descendingActivities = (activity || []).sort(
    (a, b) => b.timestamp - a.timestamp,
  );
  const activityByMonth = groupByMonth(descendingActivities).map(
    ({ month, data }) => (
      <div key={month}>
        <Subheading>{month}</Subheading>
        {data.map((event) => (
          <ActivityEvent event={event} key={JSON.stringify(event)} />
        ))}
      </div>
    ),
  );

  return (
    <>
      <Heading>Activity</Heading>

      {!activity && (
        <>
          <Placeholder width="200px" height="0.7rem" />
          <br />
          <Placeholder width="100%" height="2rem" />
        </>
      )}

      {activityByMonth}
    </>
  );
};

const mapStateToProps = (state: IState) => ({
  activity: state.activity,
});

const mapDispatchToProps = (dispatch: Dispatch) => {
  return bindActionCreators(
    {
      dispatchFetchActivities: fetchActivities,
    },
    dispatch,
  );
};

/**
 * The activity component connected to the redux store.
 */
export const ConnectedActivity = connect(
  mapStateToProps,
  mapDispatchToProps,
)(Activity);
