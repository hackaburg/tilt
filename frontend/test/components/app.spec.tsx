import { shallow } from "enzyme";
import * as React from "react";
import { UserRole } from "../../../types/roles";
import { App } from "../../src/components/app";

describe("App", () => {
  it("renders", () => {
    const boot = jest.fn();
    const app = shallow(
      <App
        settings={null}
        boot={boot}
        role={UserRole.User}
        history={{} as any}
        location={{} as any}
        match={{} as any}
        staticContext={{} as any}
      />,
    );

    expect(app).toMatchSnapshot();
  });
});
