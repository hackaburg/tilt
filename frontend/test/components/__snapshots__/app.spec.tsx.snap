// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<ThemeProvider
  theme={
    Object {
      "colorGradientEnd": "#333",
      "colorGradientStart": "#333",
      "colorLink": "#333",
      "colorLinkHover": "#333",
    }
  }
>
  <Switch>
    <Route
      component={[Function]}
      path="/login"
    />
    <Route
      component={[Function]}
      path="/verify"
    />
    <Route
      component={[Function]}
    />
  </Switch>
</ThemeProvider>
`;
